Configuration file formats for JSON, INI, YAML and TOML

infinisil opened the pull request:
###### Motivation for this change
This is a main part of the implementation of NixOS/rfcs#42 (see https://github.com/Infinisil/rfcs/blob/config-option/rfcs/0042-config-option.md#part-1-1).

Since that part of the RFC is fairly uncontroversial, I think we can merge this before the RFC is eventually accepted. This allows people to use such types in their NixOS options already.

Ping @Profpatsch @rycee @aanderse @kampka 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested that these types and generator combinations indeed can be used in modules.
- [x] Write some tests
- [x] Wrote documentation on how to use them..

Here's a simple example module though:
```nix
{ lib, config, pkgs, ... }:
let
  format = pkgs.formats.json {};
in {
  options.settings = lib.mkOption {
    type = format.type;
    default = {};
  };

  config = {
    settings.foo = 10;
    settings.bar.baz = "Hello";
    environment.etc.test.source = format.generate "config.json" config.settings;
  };
}
```

infinisil committed:
lib/generators: Extend mkValueStringDefault with float support

Infinisil commented:
Potentially/Eventually these formats might be implemented using `lazyAttrsOf` (or what becomes of it) from https://github.com/NixOS/nixpkgs/pull/70138, such that self-dependent assignments of them will be possible.

aanderse commented:
@Infinisil awesome stuff!

Profpatsch commented on lib/settings-formats.nix:
```diff
@@ -0,0 +1,51 @@
+        str
+        (attrsOf value)
+        (listOf value)
+      ]) // {
```
This is not recursive, is it?

Infinisil replied:
It is set on `value`, which is used in the definition itself, but the `description` doesn't refer to the subtypes, so it won't be infinitely long

Profpatsch replied:
Ah, missed that.

Infinisil commented:
(This still needs a bit of docs probably and a history cleanup at least btw, before anybody merges)

Infinisil commented:
I've now rebased this ontop of https://github.com/NixOS/nixpkgs/pull/70138 to implement types where values can depend on other values. E.g. this previously threw an infinite recursion error, but now it doesn't anymore:

```nix
{ lib, config, ... }:
let format = lib.formats.json;
in {
  options.settings = lib.mkOption {
    type = format.type;
    default = {};
  };

  config = {
    settings.foo = true;
    settings.bar = lib.mkIf config.settings.foo "Foo is true!";

    environment.etc.test.text = format.generate config.settings;
  };
}
```

Due to how the implementation works, it's however a bit more cumbersome if you aren't sure whether an attribute exists at all. This is how that can now be handled gracefully:
```nix
{
  # To make sure the `.name` attribute exists
  settings.name = format.empty;
  # To handle the case of `.name` being empty
  settings.greeting = "Hello, ${format.get "nobody" config.settings.name}!";
}
```

This will generate `{ greeting = "Hello, nobody!"; }`

Alternatively if you don't mind `name` being set:
```nix
{
  settings.name = lib.mkDefault "nobody";
  settings.greeting = "Hello, ${config.settings.name}!";
}
```

Which will generate `{ name = "nobody"; greeting = "Hello, nobody!"; }`

Infinisil commented on lib/settings-formats.nix:
```diff
@@ -0,0 +1,138 @@
+
+    empty = { _nixos_empty_json_value =
+      throw ("You tried to access an option value of type json that is empty. "
+        + "Use `lib.formats.json.get` on the value before using it.");
```
It might be a good idea to only have a single empty value for all

Infinisil commented:
On IRC @adisbladis brought to my attention that to support more formats like TOML, we could use `builtins.toJSON` and convert that to toml with a build-time tool, meaning the generated strings won't be available to nix at eval time anymore and the generate functions have the signature `generate :: Value -> Derivation`. This doesn't seem like a problem: Nobody should need the raw Nix strings anyways, and this will speed up evaluation too.

abbradar commented:
I have a seemingly working `toTOML` which is used for some configuration files: https://gist.github.com/abbradar/f59b55fbe3e80f1b7eab73cf10a4714e . I can check it just in case and PR it to this branch in a few days. cc @pstn

Infinisil commented:
Nice! Would be best to add it to `lib/generators.nix`

Infinisil commented:
Though I quite like the idea of doing this at build-time as mentioned above, which will also give more flexibility to change the implementation in the future. So this might not be needed for this PR specifically, but it would still be nice for other use cases in `lib.generators`.

Infinisil commented:
I now changed it to allow built-time conversions and was able to add a toml type and generator because of this. Also moved the `empty` and `get` functions outside the formats to reduce duplication. A module using this now looks like this:

```nix
{ lib, config, pkgs, ... }:
let
  format = lib.formats.toml;
in {
  options.settings = lib.mkOption {
    type = format.type;
    default = {};
  };

  config = {
    settings.foo.bar = lib.formats.empty;
    settings.bar.foo = "test${lib.formats.get "nothing" config.settings.foo.bar}";
    environment.etc.test.source = format.generate pkgs "config.toml" config.settings;
  };
}
```

Infinisil commented:
Todo: Add `__toString` to the empty value

Infinisil commented:
Todo: Change `empty` -> `unset`

Infinisil commented:
Todo: `get` -> `getOrDefault`

Mic92 commented:
I would also love to see this for home-assistant, where my configuration is getting really large, but I cannot split it into different modules properly.

danbst commented:
is this ready for merge?

The only thing I wish is `jq .`-ing files before saving them. But I guess this can be added in a separate PR as a generator `lib.formats.jsonPretty`.

Infinisil commented:
Nah, I'm in the process of writing decent tests for this. Also needs some docs still..

Nice idea though! With the generation potentially happening at build time now, using `jq` to prettify it will be very easy! I'll do that too

Profpatsch commented on lib/settings-formats.nix:
```diff
@@ -0,0 +1,132 @@
+    in valueType;
+
+    generate = pkgs: name: value: pkgs.runCommandNoCC name {
+      nativeBuildInputs = [ pkgs.jq ];
```
if you use `pkgs`, this cannot live in `lib/` anymore; anything involving build-time should like in `pkgs/build-support`.

Though maybe it could be split, so that the pure nix parts still live in `lib/`.

Profpatsch replied:
Edit: saw the `generate` structure. If we need pkgs for these functions, I’m really not sure this is a good fit for `lib/`.

Infinisil replied:
Also thought about this, but I'm not entirely sure what to do, because the `type` part doesn't need `pkgs` and it makes for sense for it to be in `lib`, unlike the `generate` part. However those two parts should be very close together, so splitting them in `pkgs` and `lib` isn't optimal either. I guess putting both in `pkgs.formats.{json,ini,yaml,toml}.{type,generate}` might be the best solution, even though I don't really like it.

danbst replied:
@Profpatsch I'd rather vote to move `pkgs/build-support/trivial-builders.nix` to `lib`. Let's leave "pure" lib to `<nix/lib>`, not `<nixpkgs/lib>`.

Infinisil replied:
@danbst I think a problem with this is that there's no reasonable default `pkgs` to use for library functions that depend on it. The only choice would be to use `import ../. {}`, however then for NixOS, that nixpkgs is imported and evaluated twice, once for `lib`, and once for `pkgs` itself (with potential overlays and stuff).

Infinisil replied:
The changes in this PR now live under `nixpkgs/pkgs/pkgs-lib`, which is intended for things that would fit into `nixpkgs/lib` but have a dependency on `pkgs`

Profpatsch commented on lib/generators.nix:
```diff
@@ -35,6 +35,7 @@ rec {
           ("generators.mkValueStringDefault: " +
            "${t} not supported: ${toPretty {} v}");
     in   if isInt      v then toString v
+    else if isFloat    v then toString v
```
nope, not a good idea. Look at the bottom of the function, there is already a case for `isFloat`.

Infinisil replied:
Oh, didn't see that. Other than Nix not printing floats with many decimals, what's the problem with this? Having this case supported here will make things a lot easier, and I don't think there's a case where it can cause harm.

Profpatsch replied:
Nix just silently removes decimals, so it’s really not an option to print them.

Infinisil replied:
Yeah but I don't think it's a problem. It seems that Nix prints 6 relevant digits, which should be enough for configurations, which is pretty much what Nix is for. I don't see how this could cause any harm.

Profpatsch replied:
Silently removing *anything* is not an option. Big :-1: on this change.

Infinisil replied:
Well I need to show floats in this PR. I guess if you're against it here I'll just have to put it somewhere else where it can be used by the functions.

Infinisil replied:
https://github.com/NixOS/nixpkgs/pull/75584/commits/a39fe137ade65c9351740f2d9792674004f995fc

Infinisil replied:
As suggested by @emilazy on IRC, it's possible to detect when an inaccurate conversion would be done. Will do that and only then throw an error

Infinisil replied:
Did the thing in the latest commits now

Profpatsch commented on lib/tests/formats.nix:
```diff
@@ -0,0 +1,186 @@
+# to run these tests:
+# nix-instantiate --eval --strict --read-write-mode nixpkgs/lib/tests/formats.nix
```
yeah, `--read-write-mode` is a good indicator that this depends on import-from-derivation.

Infinisil replied:
Yeah, I'm in the process of making it not depend on that

Infinisil replied:
Done now

Profpatsch reviewed (dismissed):
Not an in-depth review, but I have two things:

infinisil committed:
nixos/docs: Add documentation for settings options

Profpatsch commented on pkgs/pkgs-lib/default.nix:
```diff
@@ -0,0 +1,6 @@
+{ lib, pkgs }: {
```
There should be a docstring or README.md describing the idea behind pkgs-lib.

Infinisil replied:
@Profpatsch Just added a docstring explaining pkgs-lib: https://github.com/NixOS/nixpkgs/pull/75584/files#diff-fa126a2182114f76a8e8df7ed256395d. Does that look good?

infinisil committed:
lib/strings: Add floatToString

Infinisil commented:
It's probably a good idea to have a generic way to pass parameters to these formats, like how all generators in `lib.generators` take an attribute set with some options you can configure (cc @Profpatsch). E.g. the INI format has varying features across implementations: https://en.wikipedia.org/wiki/INI_file#Varying_features, all of which could be supported with additional parameters to the `lib.generators.toINI` function: https://github.com/NixOS/nixpkgs/blob/169fc2eecaf48348d1d2dc81b500fb55a94518e6/lib/generators.nix#L105-L110 (which I just discussed with @aanderse for https://github.com/NixOS/nixpkgs/pull/81940#discussion_r389391419)

If we don't support such parameters we'll probably end up with multiple different INI types over time, e.g. I know at least one config in NixOS needs [global properties](https://en.wikipedia.org/wiki/INI_file#Global_properties).

I'll take a look at implementing this.

Profpatsch commented:
> I'll take a look at implementing this.

It might make sense to have multiple types if the differences are too big, too avoid elisp-style kitchen sink functions. e.g. toINI can’t be made much more configurable without losing any sense of structure (it already assumes very little, that headers are `[` and `]` and that you need some way of escaping header contents and rendering key-value pairs).

danbst reviewed (approved):
ready for merge?

Infinisil commented:
Nope! Change of plan, see https://github.com/NixOS/rfcs/pull/42#issuecomment-599714033 and https://github.com/NixOS/nixpkgs/pull/82743, which allows for a much nicer way to implement the RFC, removing a lot of complexity from this PR.

danbst commented:
@Infinisil what's the problem to merge this first? RFC can be resolved later.

I think `unstructuredType = pkgs.formats.json.type;` would be fine for typed modules. Is API for formats going to be revamped THAT much?

Infinisil commented:
@danbst Yeah it will change a bit, specifically all the bits regarding `lazyAttrsOf` and `unset` won't be needed anymore, because with https://github.com/NixOS/nixpkgs/pull/82743, such an unstructured option recursively depending on itself shouldn't be a problem anymore (because you'd explicitly declare the option you depend on). This also has nicer semantics.

Infinisil commented:
Actually, I think this is needed in its current state after all. While with https://github.com/NixOS/nixpkgs/pull/82743 it's possible to get around infinite recursion by declaring an option, you might not always want to do that. With the `lazyAttrsOf` and `unset` bits here you can avoid infinite recursion even without declaring options for everything that depends on other things.

Profpatsch commented:
@Infinisil is this PR ready for final review?

Infinisil commented:
There's still things to do, I'll try to get to them soon. This includes:
- Adding an attrset parameter to the generators to support configuring them
- Making the INI format support the `listsAsDuplicateKeys` parameter (seen this being needed twice now)
- Finishing the docs

infinisil committed:
pkgs: Add pkgs-lib structure

infinisil committed:
pkgs-lib: Implement settings formats for JSON, INI, YAML and TOML

infinisil committed:
pkgs-lib: Add tests for formats

Infinisil commented:
Still not done, but update:
- The whole `unset` and `lazyAttrsOf` business is just damn ugly and brittle, I couldn't bring myself to leave this in. With https://github.com/NixOS/nixpkgs/pull/82743, this really won't be necessary anymore for recursive assignments.
- Formats now support passing of parameters, the `listsAsDuplicateKeys` thing is implemented for INI

Still need to finish the docs though

Mic92 commented:
We have now several copies for yaml/json types laying around in different NixOS modules. Should I just factor out the yaml type and make a new PR?

Infinisil commented:
Took some time to finish writing the docs for this. Would love somebody to proof-read or tell me how they can be improved (cc @domenkozar since I know you're pretty good at writing docs!):

![Screen Shot 2020-07-25 at 21 47 03](https://user-images.githubusercontent.com/20525370/88465111-7cbc9d80-cec0-11ea-84a8-1e99b0193898.png)

Mic92 commented on nixos/doc/manual/development/settings-options.xml:
```diff
@@ -0,0 +1,169 @@
+   <itemizedlist>
+     <listitem>
+       <para>
+         Nix-representable ones: These can trivially be mapped to a subset of Nix syntax. E.g. JSON is an example, since its values like <literal>{"foo":{"bar":10}}</literal> can be mapped directly to Nix: <literal>{ foo = { bar = 10; }; }</literal>. Other examples are INI, YAML and TOML. The following section explains the convention for these settings.
```
It's really annoying that our xml formatter does not break lines.
This is very hard to read on github.

Infinisil replied:
Oh good point. I could introduce manual line breaks, but those get annoying quickly when the text needs changes.

jtojnar replied:
This is [intentional](https://github.com/NixOS/nixpkgs/pull/68181) since hard wrapping breaks diffs. On the other hand, it is easy to enable soft wrapping using an user style. I recommend https://github.com/StylishThemes/GitHub-code-wrap.

Mic92 replied:
I see. The only alternative that would make both github code review and diffs happy is to use
break on every sentence instead of each paragraph. No idea if our xmlformatter even would have an option for that.

domenkozar replied:
Given that we're months away switching to readable documentation I wouldn't bother right now.

jtojnar commented on nixos/doc/manual/development/settings-options.xml:
```diff
@@ -0,0 +1,169 @@
+
+    # We know that `port` is assigned because we have a default value above
+    # So we can use it here without worries
+    networking.firewall.allowedTCPPorts = [ cfg.settings.port ];
```
Maybe we should avoid showing adding port to firewall whitelist since modules are not supposed to do that.

Infinisil replied:
Oh good idea! I changed it to a `user` option now.

mdlayher commented on nixos/doc/manual/development/settings-options.xml:
```diff
@@ -0,0 +1,174 @@
+   <itemizedlist>
+     <listitem>
+       <para>
+         Nix-representable ones: These can trivially be mapped to a subset of Nix syntax. E.g. JSON is an example, since its values like <literal>{"foo":{"bar":10}}</literal> can be mapped directly to Nix: <literal>{ foo = { bar = 10; }; }</literal>. Other examples are INI, YAML and TOML. The following section explains the convention for these settings.
```
Super nit: is Oxford comma (or lack of) a convention in Nix docs: e.g. "INI, YAML, and TOML."?

Infinisil replied:
I didn't even know there could be a comma there! There isn't a convention I'm fairly certain, so I'll just leave it as is for now :)

mdlayher reviewed (approved):
LGTM! I think the docs you've provided and example module are helpful. I'd be happy to be an early adopter of this package since my CoreRAD module already uses the settings attribute with a bespoke Nix -> JSON -> TOML conversion: https://github.com/NixOS/nixpkgs/pull/89781.

Profpatsch reviewed (commented):
Haven’t reviewed in-depth, but if other people are okay with it LGTM

Infinisil commented:
I think this PR is in a good state now. Unless somebody has any improvement suggestions or concerns I'll merge it this weekend!

domenkozar commented:
I'd review it if it was written with terminal length in mind, but currently it's not feasible via github ui.

roberth commented on pkgs/pkgs-lib/default.nix:
```diff
@@ -0,0 +1,11 @@
+# they depend on some packages. This notably is *not* for supporting package
+# building, instead pkgs/build-support is the place for that.
+{ lib, pkgs }: {
+  # setting format types and generators. These would fit in lib/types.nix, but
```
```suggestion
  # setting format types and generators. These do not fit in lib/types.nix, because
```

Infinisil replied:
Done with https://github.com/NixOS/nixpkgs/compare/76e8e4a26fa451bbc1832e0cd599080795227593..83b16885f526d6ab7b39e98159e2b48024f3238c#diff-fa126a2182114f76a8e8df7ed256395d

roberth commented on pkgs/pkgs-lib/default.nix:
```diff
@@ -0,0 +1,11 @@
+# pkgs-lib is for things that should be in lib, but can't beceause
```
```suggestion
# pkgs-lib is for functions that can't be in lib, because
```

Infinisil replied:
Done with adjustments in https://github.com/NixOS/nixpkgs/compare/76e8e4a26fa451bbc1832e0cd599080795227593..83b16885f526d6ab7b39e98159e2b48024f3238c#diff-fa126a2182114f76a8e8df7ed256395d

roberth commented on pkgs/top-level/stage.nix:
```diff
@@ -75,6 +75,11 @@ let
       inherit (self) runtimeShell;
     };
 
+  pkgsLib = self: super: import ../pkgs-lib {
```
An overlay is unnecessary because the same could be achieved by putting `formats` in `all-packages.nix`.

Categorizing functions separately from packages is beyond the scope of this change. It can be discussed in a separate issue/pr, although I doubt its usefulness.

Infinisil replied:
Doesn't have to be just functions though, can be arbitrary values, and other functions like `fetchgit` wouldn't fit in a pkgs-lib. I guess it won't be a problem to just put `formats` in `all-packages.nix`, but imo having a separation between "things needed to create packages" and "things that just happen to depend on pkgs" would make sense. Could still introduce that in the future though

Infinisil replied:
Not using an overlay anymore now

roberth replied:
All of `formats`, `fetchgit`, `dockerTools` depend on Nixpkgs, take arguments, produce derivations, are maintained and released in the same way. What makes `formats` so architecturally different that it needs to be treated differently?

> imo having a separation between "things needed to create packages" and "things that just happen to depend on pkgs" would make sense. Could still introduce that in the future though

If only one attribute is separate it doesn't make sense though. For it to make sense it must be applied to the whole of Nixpkgs and it must have practical benefits. The prior can be done in the future and I'm skeptical about the benefits, but would love to be proven wrong in that same future.

For now please follow the example of, say, `dockerTools`.

roberth replied:
Looks like my view of the PR was outdated (maybe caused by the current service disruption?) anyway, thanks for removing the overlay. Currently `build-support` has the role of `pkgs-lib` so you may want to move the files to `build-support/formats`. I don't really like that name either so I'm going to leave this change up to you to decide.

Infinisil replied:
The difference between e.g. `fetchgit` and `formats.json` is that `fetchgit` is needed for building many of the `pkgs.*` derivations, whereas `formats.json` doesn't have any use for any `pkgs.*`. Instead `formats` are mostly just useful for NixOS modules (really, the module system in general, so it wouldn't fit in `nixos/lib`)

This is also why I don't think `build-support` would make sense, because `formats.json` doesn't support any package building.

roberth reviewed (changes requested):
It should be noted in the docs that some of these functions may write to the store, which is not safe when Nix is handling secrets. This is in contrast to `builtins.toJSON` etc, which operate entirely in memory.

worldofpeace commented:
> I'd review it if it was written with terminal length in mind, but currently it's not feasible via github ui.

this is useful https://github.com/StylishThemes/GitHub-code-wrap

rycee commented:
An alternative that doesn't need any installations is to use the "Split" diff option. For whatever reason, it seems to wrap lines.
![github-split](https://user-images.githubusercontent.com/798147/88820004-0d48f580-d1c1-11ea-9606-2c5c6999c798.png)

Infinisil commented:
> It should be noted in the docs that some of these functions may write to the store, which is not safe when Nix is handling secrets. This is in contrast to `builtins.toJSON` etc, which operate entirely in memory.

For NixOS modules at least it doesn't make much of a difference, because the only way for a string (which `toJSON` returns) to influence a system build is by putting it in a derivation. With the interface this PR adds, this step is just done a little earlier. But yeah it can't hurt to point this out, I'll add a note

Infinisil commented on lib/generators.nix:
```diff
@@ -48,8 +48,9 @@ rec {
-    # let’s not talk about floats. There is no sensible `toString` for them.
-    else if isFloat    v then err "floats" v
+    # Floats currently can't be converted to precise strings,
+    # condition warning on nix version once this isn't a problem anymore
```
Done

Infinisil commented:
@roberth Looking good now?

Ericson2314 commented:
I wonder if `pkgs-lib` should be `pkgs/lib` to match `nixos/lib`.

Infinisil commented:
Awesome, thanks for merging!

@Ericson2314 I intentionally didn't name it `lib` because it could easily be confused with `pkgs.lib`.

fzakaria commented:
@Infinisil you might want to update this documentation section as well https://nixos.org/manual/nixpkgs/unstable/#sec-generators
It seems out of date considering the new NixOS doc you added.

fzakaria commented:
Actually reading `lib/generators.nix` it looks like it should be consolidated with the new **formats** option introduced here.
